|평가문항	|상세기준|
|:-|:-|
|1. 3가지 데이터셋의 구성을 파악하고, 데이터를 이해하는 과정이 포함되어있는가?	|feature와 label 선정을 위한 데이터 분석과정을 전개함|
|2. 3가지 데이터셋에 대해 각각 5가지 모델을 이용하여 학습을 수행하고 결과값을 얻었는가?	|모델학습 및 테스트가 정상적으로 수행되고, 결과값을 얻었음|
|3. 3가지 데이터셋에 대해 모델의 평가 지표를 선택하고, 그 이유를 근거를 바탕으로 서술하였는가?	|모델학습을 통해 얻은 결과물을 바탕으로 평가지표를 선택하고, 본인의 의견을 서술하였음|

# Mnist data

## (1) 필요한 모듈 import하기

In [72]:
import pandas as pd

import sklearn
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

import warnings
warnings.filterwarnings('ignore')

## (2) 데이터 준비

In [2]:
from sklearn.datasets import load_digits

digits = load_digits()

## (3) 데이터 이해하기

In [3]:
digits

{'data': array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
        [ 0.,  0.,  0., ..., 10.,  0.,  0.],
        [ 0.,  0.,  0., ..., 16.,  9.,  0.],
        ...,
        [ 0.,  0.,  1., ...,  6.,  0.,  0.],
        [ 0.,  0.,  2., ..., 12.,  0.,  0.],
        [ 0.,  0., 10., ..., 12.,  1.,  0.]]),
 'target': array([0, 1, 2, ..., 8, 9, 8]),
 'frame': None,
 'feature_names': ['pixel_0_0',
  'pixel_0_1',
  'pixel_0_2',
  'pixel_0_3',
  'pixel_0_4',
  'pixel_0_5',
  'pixel_0_6',
  'pixel_0_7',
  'pixel_1_0',
  'pixel_1_1',
  'pixel_1_2',
  'pixel_1_3',
  'pixel_1_4',
  'pixel_1_5',
  'pixel_1_6',
  'pixel_1_7',
  'pixel_2_0',
  'pixel_2_1',
  'pixel_2_2',
  'pixel_2_3',
  'pixel_2_4',
  'pixel_2_5',
  'pixel_2_6',
  'pixel_2_7',
  'pixel_3_0',
  'pixel_3_1',
  'pixel_3_2',
  'pixel_3_3',
  'pixel_3_4',
  'pixel_3_5',
  'pixel_3_6',
  'pixel_3_7',
  'pixel_4_0',
  'pixel_4_1',
  'pixel_4_2',
  'pixel_4_3',
  'pixel_4_4',
  'pixel_4_5',
  'pixel_4_6',
  'pixel_4_7',
  'pixel_5_0',
  'pixel_5_1',
 

In [4]:
# 객체의 메서드와 변수 확인
dir(digits)

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']

In [5]:
# 데이터셋에 담겨있는 정보 확인
digits.keys()

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

In [6]:
# keys의 data (feature)를 digits_data 변수에 저장
digits_data = digits.data

# data shape 확인
digits_data.shape

(1797, 64)

총 1797개의 행, 64개의 컬럼(feature)를 포함하고 있음  

#### [데이터 분석 목표]
각 이미지 데이터가 입력되었을 때 그 이미지가 숫자 몇을 나타내는 이미지인지 맞추는 분류 모델 학습

In [7]:
# target(lable) 데이터 분리
digits_target = digits.target

# data shape 확인
digits_target.shape

(1797,)

In [8]:
digits_target

array([0, 1, 2, ..., 8, 9, 8])

In [9]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

0부터 9까지의 데이터가 포함되어 있다.

In [10]:
digits.feature_names

['pixel_0_0',
 'pixel_0_1',
 'pixel_0_2',
 'pixel_0_3',
 'pixel_0_4',
 'pixel_0_5',
 'pixel_0_6',
 'pixel_0_7',
 'pixel_1_0',
 'pixel_1_1',
 'pixel_1_2',
 'pixel_1_3',
 'pixel_1_4',
 'pixel_1_5',
 'pixel_1_6',
 'pixel_1_7',
 'pixel_2_0',
 'pixel_2_1',
 'pixel_2_2',
 'pixel_2_3',
 'pixel_2_4',
 'pixel_2_5',
 'pixel_2_6',
 'pixel_2_7',
 'pixel_3_0',
 'pixel_3_1',
 'pixel_3_2',
 'pixel_3_3',
 'pixel_3_4',
 'pixel_3_5',
 'pixel_3_6',
 'pixel_3_7',
 'pixel_4_0',
 'pixel_4_1',
 'pixel_4_2',
 'pixel_4_3',
 'pixel_4_4',
 'pixel_4_5',
 'pixel_4_6',
 'pixel_4_7',
 'pixel_5_0',
 'pixel_5_1',
 'pixel_5_2',
 'pixel_5_3',
 'pixel_5_4',
 'pixel_5_5',
 'pixel_5_6',
 'pixel_5_7',
 'pixel_6_0',
 'pixel_6_1',
 'pixel_6_2',
 'pixel_6_3',
 'pixel_6_4',
 'pixel_6_5',
 'pixel_6_6',
 'pixel_6_7',
 'pixel_7_0',
 'pixel_7_1',
 'pixel_7_2',
 'pixel_7_3',
 'pixel_7_4',
 'pixel_7_5',
 'pixel_7_6',
 'pixel_7_7']

64 = 8 x 8 픽셀 이미지가 담겨있음

In [11]:
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

## (4) train, test 데이터 분리

In [12]:
type(digits_data)

numpy.ndarray

In [13]:
# ndarray digits_data를 dataframe으로 변경
df_m = pd.DataFrame(digits_data, columns = digits.feature_names)
df_m

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        5.0        0.0   
3           0.0        0.0        0.0        8.0  ...        9.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0           0.0        0.0        6.0       13.0       10.0        0.0   
1           0.0        0.0        0.0       11.0       16.0       10.0   
2           0.0        0.0        0.0        3.0       11.0       16.0   
3           0.0        0.0        7.0       13.0       13.0        9.0   
4           0.0        0.0        0.0        2.0       16.0        4.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
0           0.0        0.0  
1           0.0        0.0  
2           9.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
...         ...        ...  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[1797 rows x 64 columns]

In [14]:
# target (label) 데이터 추가
df_m['label'] = digits.target
df_m

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_7  pixel_7_0  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        0.0        0.0   
3           0.0        0.0        0.0        8.0  ...        0.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        0.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        0.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        0.0        0.0   

      pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  pixel_7_6  \
0           0.0        6.0       13.0       10.0        0.0        0.0   
1           0.0        0.0       11.0       16.0       10.0        0.0   
2           0.0        0.0        3.0       11.0       16.0        9.0   
3           0.0        7.0       13.0       13.0        9.0        0.0   
4           0.0        0.0        2.0       16.0        4.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        2.0       14.0       15.0        9.0        0.0   
1793        0.0        6.0       16.0       14.0        6.0        0.0   
1794        0.0        2.0        9.0       13.0        6.0        0.0   
1795        0.0        5.0       12.0       16.0       12.0        0.0   
1796        1.0        8.0       12.0       14.0       12.0        1.0   

      pixel_7_7  label  
0           0.0      0  
1           0.0      1  
2           0.0      2  
3           0.0      3  
4           0.0      4  
...         ...    ...  
1792        0.0      9  
1793        0.0      0  
1794        0.0      8  
1795        0.0      9  
1796        0.0      8  

[1797 rows x 65 columns]

### (4.1) Data splitting

In [15]:
X_train, X_test, y_train, y_test = train_test_split(digits_data,
                                                    digits_target,
                                                    test_size = 0.2,
                                                    random_state = 2024)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

print("\ny_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (1437, 64)
X_test shape: (360, 64)

y_train shape: (1437,)
y_test shape: (360,)


## (5) 다양한 모델로 학습시켜보기

### (5.1) Decision Tree

In [16]:
# 의사결정나무 불러오기
decision_tree = DecisionTreeClassifier(random_state = 32)

# 학습
decision_tree.fit(X_train, y_train)

DecisionTreeClassifier(random_state=32)

In [17]:
# 평가
y_pred = decision_tree.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      1.00      0.93        31
           1       0.91      0.78      0.84        41
           2       0.87      0.94      0.90        35
           3       0.79      0.94      0.86        33
           4       0.88      0.91      0.90        33
           5       0.94      0.97      0.95        31
           6       0.97      0.85      0.91        41
           7       0.94      0.85      0.89        39
           8       0.81      0.87      0.84        39
           9       0.88      0.78      0.83        37

    accuracy                           0.88       360
   macro avg       0.89      0.89      0.88       360
weighted avg       0.89      0.88      0.88       360



정확도는 0.88이며, 5를 가장 잘 예측한다 (precision, recall 모두 0.9 이상).

### (5.2) Random Forest

In [18]:
# 랜덤포레스트 불러오기
random_forest = RandomForestClassifier(random_state = 32)

# 학습
random_forest.fit(X_train, y_train)

RandomForestClassifier(random_state=32)

In [19]:
# 평가
y_pred = random_forest.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        31
           1       1.00      1.00      1.00        41
           2       1.00      1.00      1.00        35
           3       0.94      1.00      0.97        33
           4       0.97      1.00      0.99        33
           5       0.97      0.97      0.97        31
           6       1.00      0.98      0.99        41
           7       0.93      1.00      0.96        39
           8       0.97      0.97      0.97        39
           9       0.97      0.84      0.90        37

    accuracy                           0.97       360
   macro avg       0.98      0.98      0.97       360
weighted avg       0.98      0.97      0.97       360



정확도가 0.97로 올랐으며, 0, 1, 2는 완벽하게 예측했다.  
상대적으로 9를 가장 잘 예측하지 못함.

### (5.3) SVM

In [20]:
# svm 불러오기
svm_model = svm.SVC(random_state = 32)

# 학습
svm_model.fit(X_train, y_train)

SVC(random_state=32)

In [21]:
# 평가
y_pred = svm_model.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        31
           1       1.00      1.00      1.00        41
           2       0.97      1.00      0.99        35
           3       1.00      0.97      0.98        33
           4       1.00      1.00      1.00        33
           5       1.00      0.97      0.98        31
           6       1.00      1.00      1.00        41
           7       0.97      1.00      0.99        39
           8       0.97      1.00      0.99        39
           9       0.97      0.95      0.96        37

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



정확도가 0.99로 올랐으며, 0, 1, 6는 완벽하게 예측했다.  
사실상 모든 숫자를 잘 예측한다.  
어디까지나 상대적으로 9를 가장 잘 예측하지 못함.

### (5.4) SGD

In [22]:
# sgd 불러오기
sgd_model = SGDClassifier(random_state = 32)

# 학습
sgd_model.fit(X_train, y_train)

SGDClassifier(random_state=32)

In [23]:
# 평가
y_pred = sgd_model.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      1.00      0.98        31
           1       0.95      0.90      0.92        41
           2       0.95      1.00      0.97        35
           3       0.97      0.94      0.95        33
           4       1.00      1.00      1.00        33
           5       1.00      0.94      0.97        31
           6       1.00      0.98      0.99        41
           7       0.97      1.00      0.99        39
           8       0.82      0.92      0.87        39
           9       0.97      0.89      0.93        37

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.96      0.96      0.96       360



정확도는 0.96이었으며, 4는 완벽하게 예측했다.  
상대적으로 8을 가장 잘 예측하지 못함.

### (5.5) Logistic Regression

In [24]:
# logistic regression 불러오기
logistic_model = LogisticRegression(random_state = 32)

# 학습
logistic_model.fit(X_train, y_train)

LogisticRegression(random_state=32)

In [25]:
# 평가
y_pred = logistic_model.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        31
           1       0.95      0.93      0.94        41
           2       0.92      1.00      0.96        35
           3       0.97      0.94      0.95        33
           4       0.97      1.00      0.99        33
           5       0.93      0.90      0.92        31
           6       1.00      0.98      0.99        41
           7       0.97      1.00      0.99        39
           8       0.90      0.90      0.90        39
           9       0.94      0.92      0.93        37

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.96      0.96      0.96       360



정확도가 0.96이었으며, 0은 완벽하게 예측했다.  
상대적으로 8을 가장 잘 예측하지 못함.

## (6) 결론  
Mnist 데이터에서는 (하이퍼 파라미터 튜닝이 없을 때) SVM 모델이 정확도, 정밀도, 재현율, F1-점수에서 모든 모델 중 가장 높은 성능을 보였다.

# Wine

## (1) 필요한 모듈 import하기  
위에서 불러왔으므로 생략

## (2) 데이터 준비

In [26]:
from sklearn.datasets import load_wine

wine = load_wine()

## (3) 데이터 이해하기

In [27]:
# 객체의 메서드와 변수 확인
dir(wine)

['DESCR', 'data', 'feature_names', 'frame', 'target', 'target_names']

In [28]:
# 데이터셋에 담겨있는 정보 확인
wine.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

In [29]:
# keys의 data (feature)를 cancer_data 변수에 저장
wine_data = wine.data

# data shape 확인
wine_data.shape

(178, 13)

총 178개의 행, 13개의 컬럼(feature)를 포함하고 있음  

#### [데이터 분석 목표]
와인을 13가지 화학적 분석 기반으로 와인을 분류.

In [30]:
# target(lable) 데이터 분리
wine_target = wine.target

# data shape 확인
wine_target.shape

(178,)

In [31]:
wine_target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

In [32]:
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

세 가지 클래스 (class_0, class_1, class_2)를 가짐

In [33]:
wine.feature_names

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

30가지의 화학적 특성이 포함되어 있다.

In [34]:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            10.6  30.0    19.5   3.3
    Ma

## (4) train, test 데이터 분리

In [35]:
type(wine_data)

numpy.ndarray

In [36]:
# ndarray cance_data를 dataframe으로 변경
df_w = pd.DataFrame(wine_data, columns = wine.feature_names)
df_w

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

In [37]:
# target (label) 데이터 추가
df_w['label'] = wine.target
df_w

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  label  
0                            3.92   1065.0      0  
1                            3.40   1050.0      0  
2                            3.17   1185.0      0  
3                            3.45   1480.0      0  
4                            2.93    735.0      0  
..                            ...      ...    ...  
173                          1.74    740.0      2  
174                          1.56    750.0      2  
175                          1.56    835.0      2  
176                          1.62    840.0      2  
177                          1.60    560.0      2  

[178 rows x 14 columns]

### (4.1) Data splitting

In [38]:
X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                    wine_target,
                                                    test_size = 0.2,
                                                    random_state = 2024)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

print("\ny_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (142, 13)
X_test shape: (36, 13)

y_train shape: (142,)
y_test shape: (36,)


## (5) 다양한 모델로 학습시켜보기

### (5.1) Decision Tree

In [39]:
# 의사결정나무 불러오기
decision_tree = DecisionTreeClassifier(random_state = 32)

# 학습
decision_tree.fit(X_train, y_train)

DecisionTreeClassifier(random_state=32)

In [40]:
# 평가
y_pred = decision_tree.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        11
           1       0.69      0.69      0.69        13
           2       0.67      0.67      0.67        12

    accuracy                           0.78        36
   macro avg       0.79      0.79      0.79        36
weighted avg       0.78      0.78      0.78        36



정확도는 0.78이며, class_0은 완벽하게 예측했다.  
모든 class에 대해 모든 평가 지표가 과반 이상이다.

### (5.2) Random Forest

In [41]:
# 랜덤포레스트 불러오기
random_forest = RandomForestClassifier(random_state = 32)

# 학습
random_forest.fit(X_train, y_train)

RandomForestClassifier(random_state=32)

In [42]:
# 평가
y_pred = random_forest.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        11
           1       1.00      1.00      1.00        13
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



모든 class에 대해 완벽하게 예측했다.

### (5.3) SVM

In [43]:
# svm 불러오기
svm_model = svm.SVC(random_state = 32)

# 학습
svm_model.fit(X_train, y_train)

SVC(random_state=32)

In [44]:
# 평가
y_pred = svm_model.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.90      0.82      0.86        11
           1       0.53      0.69      0.60        13
           2       0.44      0.33      0.38        12

    accuracy                           0.61        36
   macro avg       0.62      0.61      0.61        36
weighted avg       0.61      0.61      0.61        36



정확도는 0.61이며, class_0을 가장 잘 예측했고, class_2를 가장 잘 예측하지 못했다.  
class_2에 대해 모든 평가 지표가 절반 미만이다.

### (5.4) SGD

In [45]:
# sgd 불러오기
sgd_model = SGDClassifier(random_state = 32)

# 학습
sgd_model.fit(X_train, y_train)

SGDClassifier(random_state=32)

In [46]:
# 평가
y_pred = sgd_model.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.73      0.84        11
           1       0.53      0.69      0.60        13
           2       0.45      0.42      0.43        12

    accuracy                           0.61        36
   macro avg       0.66      0.61      0.63        36
weighted avg       0.65      0.61      0.62        36



정확도는 0.61이며, class_0을 가장 잘 예측한다. SVM모델에 비해서는 class_2를 더 잘 예측한다.  
class_2에 대해 모든 평가 지표가 절반 미만이다.

### (5.5) Logistic Regression

In [47]:
# logistic regression 불러오기
logistic_model = LogisticRegression(random_state = 32)

# 학습
logistic_model.fit(X_train, y_train)

LogisticRegression(random_state=32)

In [48]:
# 평가
y_pred = logistic_model.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.91      0.91      0.91        11
           1       0.86      0.92      0.89        13
           2       1.00      0.92      0.96        12

    accuracy                           0.92        36
   macro avg       0.92      0.92      0.92        36
weighted avg       0.92      0.92      0.92        36



Decision Tree 모델에 비해 모든 class에 대해 잘 예측했다.

## (6) 결론
와인 데이터에 대해서 (하이퍼파라미터 튜닝이 없을 때) 랜덤 포레스트 모델이 완벽한 성능을 나타냈다.  
로지스틱 회귀 모델도 높은 정확도와 높은 평균 성능을 나타낸다.  
SVM와 SGD 모델은 (하이퍼파라미터 튜닝이 없을 때) 와인 데이터에 대한 성능이 낮다.

# Breast cancer

## (1) 필요한 모듈 import하기  
위에서 불러왔으므로 생략

## (2) 데이터 준비

In [49]:
from sklearn.datasets import load_breast_cancer

cancer = load_breast_cancer()

## (3) 데이터 이해하기

In [50]:
# 객체의 메서드와 변수 확인
dir(cancer)

['DESCR',
 'data',
 'data_module',
 'feature_names',
 'filename',
 'frame',
 'target',
 'target_names']

In [51]:
# 데이터셋에 담겨있는 정보 확인
cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [52]:
# keys의 data (feature)를 cancer_data 변수에 저장
cancer_data = cancer.data

# data shape 확인
cancer_data.shape

(569, 30)

총 569개의 행, 30개의 컬럼(feature)를 포함하고 있음  

#### [데이터 분석 목표]
종양을 30가지 clinical 측정값을 기반으로 종양이 양성인지 악성인지를 분류함.

In [53]:
# target(lable) 데이터 분리
cancer_target = cancer.target

# data shape 확인
cancer_target.shape

(569,)

In [54]:
cancer_target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

In [55]:
cancer.target_names

array(['malignant', 'benign'], dtype='<U9')

악성종양(malignant), 양성종양(benign)으로 구분  
- 양성종양(benign): 서서히 성장하며, 신체 여러 부위에 전이하지도 않아 __생명에 위험을 초래하지 않는 경우가 대부분__  
- 악성종양(malignant): 빠른 성장과 확산, 전이로 인해 __생명의 위험을 초래__

In [56]:
cancer.feature_names

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

30가지의 화학적 특성이 포함되어 있다.

In [57]:
print(cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

## (4) train, test 데이터 분리

In [58]:
type(cancer_data)

numpy.ndarray

In [59]:
# ndarray cance_data를 dataframe으로 변경
df_c = pd.DataFrame(cancer_data, columns = cancer.feature_names)
df_c

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

In [60]:
# target (label) 데이터 추가
df_c['label'] = cancer.target
df_c

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  label  
0                  0.2654          0.4601                  0.11890      0  
1                  0.1860          0.2750                  0.08902      0  
2                  0.2430          0.3613                  0.08758      0  
3                  0.2575          0.6638                  0.17300      0  
4                  0.1625          0.2364                  0.07678   

### (4.1) Data splitting

In [61]:
X_train, X_test, y_train, y_test = train_test_split(cancer_data,
                                                    cancer_target,
                                                    test_size = 0.2,
                                                    random_state = 2024)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

print("\ny_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (455, 30)
X_test shape: (114, 30)

y_train shape: (455,)
y_test shape: (114,)


## (5) 다양한 모델로 학습시켜보기

### (5.1) Decision Tree

In [62]:
# 의사결정나무 불러오기
decision_tree = DecisionTreeClassifier(random_state = 32)

# 학습
decision_tree.fit(X_train, y_train)

DecisionTreeClassifier(random_state=32)

In [63]:
# 평가
y_pred = decision_tree.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.96      0.94        46
           1       0.97      0.94      0.96        68

    accuracy                           0.95       114
   macro avg       0.94      0.95      0.95       114
weighted avg       0.95      0.95      0.95       114



정확도는 0.95이며, 재현율도 매우 높은 편이다.  
암을 판단하는 데에는 FN가 가장 치명적이므로 재현율이 특히 중요하다.

### (5.2) Random Forest

In [64]:
# 랜덤포레스트 불러오기
random_forest = RandomForestClassifier(random_state = 32)

# 학습
random_forest.fit(X_train, y_train)

RandomForestClassifier(random_state=32)

In [65]:
# 평가
y_pred = random_forest.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.98      0.91      0.94        46
           1       0.94      0.99      0.96        68

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114



정확도는 0.96이며, Decision Tree에 0(악성종양)에 대한 재현율이 상대적으로 약간 낮다.  
암을 판단하는 데에는 FN가 가장 치명적이므로 재현율이 특히 중요하다.

### (5.3) SVM

In [66]:
# svm 불러오기
svm_model = svm.SVC(random_state = 32)

# 학습
svm_model.fit(X_train, y_train)

SVC(random_state=32)

In [67]:
# 평가
y_pred = svm_model.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.76      0.86        46
           1       0.86      1.00      0.93        68

    accuracy                           0.90       114
   macro avg       0.93      0.88      0.89       114
weighted avg       0.92      0.90      0.90       114



정확도는 0.90이며, 악성종양에 대한 재현율이 다소 낮다.  
암을 판단하는 데에는 FN가 가장 치명적이므로 재현율이 특히 중요하다.

### (5.4) SGD

In [68]:
# sgd 불러오기
sgd_model = SGDClassifier(random_state = 32)

# 학습
sgd_model.fit(X_train, y_train)

SGDClassifier(random_state=32)

In [69]:
# 평가
y_pred = sgd_model.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.61      0.76        46
           1       0.79      1.00      0.88        68

    accuracy                           0.84       114
   macro avg       0.90      0.80      0.82       114
weighted avg       0.88      0.84      0.83       114



정확도는 0.84이며, 악성종양에 대한 재현율이 상당히 낮다.  
암을 판단하는 데에는 FN가 가장 치명적이므로 재현율이 특히 중요하다.

### (5.5) Logistic Regression

In [70]:
# logistic regression 불러오기
logistic_model = LogisticRegression(random_state = 32)

# 학습
logistic_model.fit(X_train, y_train)

LogisticRegression(random_state=32)

In [71]:
# 평가
y_pred = logistic_model.predict(X_test)

# 평가 지표 확인
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.89      0.91      0.90        46
           1       0.94      0.93      0.93        68

    accuracy                           0.92       114
   macro avg       0.92      0.92      0.92       114
weighted avg       0.92      0.92      0.92       114



정확도는 0.92이며, SVM과 SGD에 비해 악성종양에 대한 재현율이 높다.
암을 판단하는 데에는 FN가 가장 치명적이므로 재현율이 특히 중요하다.

## (6) 결론  
- (하이퍼파라미터 튜닝이 없을 때) 유방암 데이터에 대해 랜덤 포레스트 모델이 가장 높은 정확도를 보였으며, 특히 양성 종양 예측 능력이 매우 우수함.
- 의자결정나무와 로지스틱 회위 모델도 높은 정확도를 보여주며, 효과적인 분류 성능을 제공함.
- (하이퍼파라미터 튜닝이 없을 때) SVM과 SGD 모델은 악성 종양에 대한 재현율이 낮다.